pkg/specgen/generate/container_create.go

```diff
@@ -199,20 +203,20 @@ func ExecuteCreate(ctx context.Context, rt *libpod.Runtime, runtimeSpec *spec.Sp
 	return ctr, rt.PrepareVolumeOnCreateContainer(ctx, ctr)
 }
 
-func ExtractCDIDevices(s *specgen.SpecGenerator) []libpod.CtrCreateOption {
+// ExtractCDIDevices process the list of Devices in the spec and determines if any of these are CDI devices.
+// The specified CDI device database is used to check the validity of the devices.
+func ExtractCDIDevices(s *specgen.SpecGenerator, cdiDeviceDB cdi.RegistryDeviceDB) []libpod.CtrCreateOption {
 	devs := make([]spec.LinuxDevice, 0, len(s.Devices))
 	var cdiDevs []string
 	var options []libpod.CtrCreateOption
 
 	for _, device := range s.Devices {
-		isCDIDevice, err := cdi.HasDevice(device.Path)
-		if err != nil {
-			logrus.Debugf("CDI HasDevice Error: %v", err)
-		}
-		if err == nil && isCDIDevice {
+		if d := cdiDeviceDB.GetDevice(device.Path); d != nil {
```

Comment:
nitpick: you could also do syntactic filtering/check for CDI device references here (using `IsQualifiedName`).

One benefit of that would be that unresolvable CDI devices would go through the CDI resolution/injection attempt and come back as unresolved. This would potentially allow you to produce better/more contextually accurate error messages. AFAICT, with the current implementation such devices/errors are only caught later in the container creation pipeline, probably only in the runtime once it tries to inject the device into the container.

Reply:
I wanted to avoid relying on `IsQualifiedName` here. One motivator is that I want to have some flexibility (in our library) for determining whether or not a device is a CDI device without requiring that clients such as podman be updated to support these changes. I understand that they would require an update, but this would be limited to updating a dependency.

Another motivation is that a fully-qualified device name does not mean that the device has an associated CDI spec.

|  | example | fully qualified | has CDI spec | action at filtering | result at injection | 
|---|---|---------------|--------------|-------------|-------------------|
| 1 | `nvidia.com/gpu=invalid` | yes |  no | assumed to be standard device | errors with `Error: stat nvidia.com/gpu=invalid: no such file or directory` |
| 2 | `nvidia.com/gpu=gpu0` | yes |  yes | identified as CDI device | resolves and injects modifications for `nvidia.com/gpu=gpu0` |
| 3 | `gpu0` | no | yes |  assumed to be standard device (CDI currently requires a fully-qualified name) | errors with `Error: stat gpu0: no such file or directory` | 
| 4 | `/dev/nvidia0` | no | no |  assumed to be standard device | injects standard device node `/dev/nvidia0` | 

Using `IsQualifiedName` would change the filtering behaviour for 1 and would result in a *warning* and the container being launched without injecting the device. Is this better behaviour than exiting with the error shown? Is it more desirable to have the contains start with no modifications to the spec than to error out for an invalid device?

Thinking about this a bit more, I think raising an error for unresolved CDI devices instead of just issuing a warning would be more in keeping with the current behaviour.

Reply:
That's not how it would behave IMO. If device resolution/injection fails for any of the devices, inject will also return a non-nil error.

Reply:
Ah. Good point. I just confirmed that `InjectDevices` also returns an error if `unresolved` is non-empty meaning that injection will fail for any unresolved devices.

Reply:
@klihub I have reworked the implementation to use `IsQualifiedName` as discussed. I abstracted this behind a function `isCDIDevice` which could later be moved into our API if this makes sense.

This also removes the requirement that the registry be instantiated here meaning that the error presented to the user should be clearer. When selecting a fully-qualified but unresolvable CDI device we now see:

```
sudo ./bin/podman run -ti --rm  --net=host --device=nvidia.com/gpu=gpu3 ubuntu:18.04 nvi
dia-smi -L
WARN[0000] Ignoring global metacopy option, not supported with booted kernel
WARN[0000] Could not resolve the following CDI devices: [nvidia.com/gpu=gpu3]
Error: error setting up CDI devices: unresolvable CDI devices nvidia.com/gpu=gpu3
```

If there were errors creating the registry these are also shown as a warning (in the event of resolution errors).